server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,194 @@
+
+   public BLPOP() {
+      super(-3, 1, -2, 1);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
+         ChannelHandlerContext ctx,
+         List<byte[]> arguments) {
+      EmbeddedMultimapListCache<byte[], byte[]> listMultimap = handler.getListMultimap();
+      var lastKeyIdx = arguments.size() - 1;
+      var filterKeys = arguments.subList(0, lastKeyIdx);
+      // Using last arg as timeout if it can be a double
+      var argTimeout = ArgumentUtils.toDouble(arguments.get(lastKeyIdx));
+      if (argTimeout < 0) {
+         RespErrorUtil.mustBePositive(handler.allocator());
+         return handler.myStage();
+      }
+      long timeout = (long) (argTimeout * Duration.ofSeconds(1).toMillis());
+
+      // If all the keys are empty or null, create a listener for each key
+      // otherwise return the left value of the first non empty list
+      var pollStage = pollAllKeys(listMultimap, arguments);
+      // If no value returned, we need subscribers
+      return handler.stageToReturn(pollStage.thenCompose(v -> {
+         // addSubscriber call can rise exception that needs to be reported
+         // as error
+         var retStage = (v != null && !v.isEmpty())
+               ? CompletableFuture.completedFuture(v)
+               : addSubscriber(listMultimap, filterKeys, timeout, handler, ctx);
+         return retStage;
+      }), ctx, Consumers.COLLECTION_BULK_BICONSUMER);
+   }
+
+   CompletionStage<Collection<byte[]>> pollKeyValue(EmbeddedMultimapListCache<byte[], byte[]> mmList, byte[] key) {
+      return mmList.pollFirst(key, 1)
+            .thenApply((v) -> (v == null || v.isEmpty())
+                  ? null
+                  : Arrays.asList(key, v.iterator().next()));
+   }
+
+   CompletionStage<Collection<byte[]>> addSubscriber(EmbeddedMultimapListCache<byte[], byte[]> listMultimap,
+         List<byte[]> filterKeys, long timeout, Resp3Handler handler,
+         ChannelHandlerContext ctx) {
+      if (log.isTraceEnabled()) {
+         log.tracef("Subscriber for keys: " +
+               filterKeys.toString());
+      }
+      AdvancedCache<Object, Object> cache = handler.cache().withMediaType(MediaType.APPLICATION_OCTET_STREAM, null);
+      DataConversion vc = cache.getValueDataConversion();
+      PubSubListener pubSubListener = new PubSubListener(ctx.channel());
+      EventListenerKeysFilter filter = new EventListenerKeysFilter(filterKeys.toArray(byte[][]::new));
+      CacheEventConverter<Object, Object, Object> converter = (key, oldValue, oldMetadata, newValue, newMetadata,
+            eventType) -> vc.fromStorage(newValue);
+      CompletionStage<Void> addListenerStage = cache.addListenerAsync(pubSubListener, filter, converter);
+      final var pubSubFuture = pubSubListener.getFuture();
+      return CompletionStages.handleAndCompose(addListenerStage, (ignore, t) -> {
```

Comment:
It doesn't appear we need all of the chained Compose calls. All you need to do is just `return pubSubFuture` as far as I can see as the last line.